### PR TITLE
#10220 - Use UIScreen.maximumFramesPerSecond for dynamic slow frame threshold 

### PR DESCRIPTION
**Pull request by JesusRojass**
### Discussion

Fix for #10220 by deriving slow-frame budget from UIScreen.maximumFramesPerSecond

User Statements:
- Slow rendering values on tvOS approximately 150 times higher (~60%) than on iOS (<1%)
- ProMotion devices reporting inaccurate frame metrics during refresh rate changes
- 50Hz tvOS displays (Europe) incorrectly flagging smooth 18ms frames as slow
- Static 60Hz assumption breaks on modern Apple hardware with variable refresh rates

What was done?

- Replace fixed 16 ms with slowBudget = 1.0 / UIScreen.mainScreen.maximumFramesPerSecond (seconds). Keeps frozen at 700 ms. Behavior at 60 Hz preserved.  
- Cache maxFPS and recompute slowBudget only when needed.
- Scoped refresh of the cache to tvOS on app-did-become-active.

### Testing

All green ✅ 

Added tests:
- 50 fps budget ~20 ms, 60 fps ~16.67 ms, 120 fps ~8.33 ms - assert slow/ok classification and frozen remains at 700 ms.
- tvOS-specific test mirroring testSlowFrameIsRecorded.
- Tests stub maximumFramesPerSecond locally to avoid global state.

### API Changes

No API Changes

### CODE REVIEW

File: FirebasePerformance/Tests/Unit/FPRScreenTraceTrackerTest.m [hunk 1]
```diff
@@ -895,6 +898,242 @@ - (void)testScreenTracesAreCreatedForContainerViewControllerSubclasses {
...
+                   finalSlowFramesCount, newSlowFramesCount,
+                   @"Frame at 19ms should NOT be marked as slow at 50 FPS (20ms threshold)");
+             }];
+}
+#endif
+
+/** Tests that the epsilon value correctly handles edge cases around 59.94 vs 60 Hz displays.
+ *  Frames right at the threshold should not be miscounted due to floating point precision.
+ */
+- (void)testSlowFrameEpsilonHandlesBoundaryCases {
```
Comment (eBlender) [inline FirebasePerformance/Tests/Unit/FPRScreenTraceTrackerTest.m]: nit: I think we could rename this to something that improves readability of what the test is doing. 
e.g "testSlowFrameRate_isHandled_inEdgeCases"

File: FirebasePerformance/Sources/AppActivity/FPRScreenTraceTracker.m [hunk 2]
```diff
@@ -112,6 +125,25 @@ - (instancetype)init {
     atomic_store_explicit(&_totalFramesCount, 0, memory_order_relaxed);
     atomic_store_explicit(&_frozenFramesCount, 0, memory_order_relaxed);
     atomic_store_explicit(&_slowFramesCount, 0, memory_order_relaxed);
+
+    // Initialize cached values with defaults. These will be updated by updateCachedSlowBudget,
+    // but having defaults ensures reasonable behavior if initialization is delayed or fails.
+    // Use 59 FPS as default to match legacy behavior (60 FPS devices use 59 FPS threshold).
+    _cachedMaxFPS = 59;
```
Comment (visumickey) [inline FirebasePerformance/Sources/AppActivity/FPRScreenTraceTracker.m]: Why are using 59 instead of 60? I thought we should be having the threshold here, right? Why do we hold it at 59?

Either ways, make it a constant and use the constant in the code. This value is being used here and down below. Let's try to avoid using a constant value across multiple parts of the code.

File: FirebasePerformance/Sources/AppActivity/FPRScreenTraceTracker.m [hunk 3]
```diff
@@ -183,13 +232,50 @@ - (void)appWillResignActiveNotification:(NSNotification *)notification {
...
+  UIScreen *mainScreen = [UIScreen mainScreen];
+  NSInteger maxFPS = 0;
+  if (mainScreen) {
+    maxFPS = mainScreen.maximumFramesPerSecond;
+  }
+  if (maxFPS > 0) {
+    _cachedMaxFPS = maxFPS;
+    // Preserve legacy behavior: 60 FPS devices historically used 59 FPS threshold
+    // to avoid too many false positives for slow frames.
+    NSInteger effectiveFPS = (maxFPS == 60) ? 59 : maxFPS;
```
Comment (visumickey) [inline FirebasePerformance/Sources/AppActivity/FPRScreenTraceTracker.m]: This also needs to change.

Comment (JesusRojass) [reply FirebasePerformance/Sources/AppActivity/FPRScreenTraceTracker.m]: Missed this one, sorry

Comment (JesusRojass) [reply FirebasePerformance/Sources/AppActivity/FPRScreenTraceTracker.m]: Solved

**Comment by JesusRojass** [comment]
@visumickey Done!

**Comment by visumickey** [comment]
The failure here is non blocking. So, merging the changes.

**Comment by ncooke3** [comment]
Hi @JesusRojass , could you please update the performance changelog to include this fix? The changelog is in performance/CHANGELOG.md, and you can use `# Unreleased` as the header title. Please include the issue number in the note. Example:

```
# Unreleased
- [fixed] <brief note>. (#10220)
```

**Comment by JesusRojass** [comment]
@ncooke3 Will do ASAP, I knew something was missing 😅
